### DISCUSSION TITLE
How to make bot react to users' mentions?

### DISCUSSION THREAD

**Discussion by Martyn0324** [Rapptz/discord.py#5877] [Q&A] [answered]
I want to make a minigame using my Discordpy bot. The idea is:

1. User starts minigame with a command;
2. Bot confirms that the minigame have been started;
3. Bot asks for a 2nd player to play the minigame;
4. User mentions 2nd player with mention(@User in discord);
5. Minigame starts.

My only problem in this case is the 4th stage: I can't make my bot detect mentions when the user answers its question.

I've tried 2 different codes based on what I found at the internet:

### First Code:
```
@client.command()
async def duel1(ctx):
    await ctx.send("Cráááá! {0.author.mention}, so you desire to wash your honor in blood and pleasure yourself with revenge's sweet taste? Cráááá! Very well...".format(ctx))
    await ctx.send('https://media1.tenor.com/images/ce1b9df56301c8443bcb4b5edbbfbdf7/tenor.gif?itemid=4581114')
    await ctx.send("Then select your opponent! Cráááá!")
    #From this point, the bot won't do anything more.
    #The following is the code used to make the bot react to a message that contains a mention.
    async def opponent(msg, member: discord.Member=None):
        if not member:
            await msg.send('Cráááááá! You need to ping your enemy! Are you scared? Cráááá!')
            return
        if member:
            await msg.message.channel.send("Cráááá! {0.member.mention}, you have been challenged by {0.author.mention} to a duel! As the challenged one, you can be the first to choose the weapon that you're gonna use during this duel. Crááá!".format(msg))
```

### Second Code:
```
@client.command()
async def duel2(ctx):
    await ctx.send("Cráááá! {0.author.mention}, so you desire to wash your honor in blood and pleasure yourself with revenge's sweet taste? Cráááá! Very well...".format(ctx))
    await ctx.send('https://media1.tenor.com/images/ce1b9df56301c8443bcb4b5edbbfbdf7/tenor.gif?itemid=4581114')
    await ctx.send("Then select your oponent! Cráááá!")
    def check(message):
        return message.author == ctx.author and message.content == str(discord.mentions)
    msg = await client.wait_for('message', timeout=20, check = check)
    if asyncio.TimeoutError and msg is None:
        await ctx.send("Cráááááá! What's the problem? Are you scared? Crááá! Call me when you stop crying out of fear! Cráááá!")
    elif msg:
        await ctx.send("Cráááá! {0.user.mention} you have been defied by {0.author.mention} to a duel! As the challenged one, you can be the first to choose the weapon that you're gonna use during this duel. Crááá!".format(ctx))
```
I've managed to make the TimeoutError work. However, as I've tried to correct the code, I messed up with it and now not even the Timeout is working. The situation is the same as the first code: after `await ctx.send("Then select your opponent! Cráááá!")` , the bot won't say nor react to anything.

PS: I do not want to use users IDs or databases in order to make the bot detect mentions.

Links I've checked:
[https://discordpy.readthedocs.io/en/latest/api.html?highlight=wait_for#discord.Client.wait_for](url)
[https://discordpy.readthedocs.io/en/latest/api.html?highlight=user%20mention#discord.ClientUser.mention](url)
[https://discordpy.readthedocs.io/en/latest/api.html?highlight=discord%20mentions#discord.Message.raw_mentions](url)
[https://github.com/Rapptz/discord.py/blob/async/examples/reply.py#L12](url)

**Comment by Rapptz** [comment, marked as answer]
Your second code is closer to what you want to do, however the check function is wrong. There is no such thing as `discord.mentions` and taking the `str` out of it makes little sense.

Messages have a [`Message.mentions`](https://discordpy.readthedocs.io/en/latest/api.html#discord.Message.mentions) attribute that allow you to get the members being mentioned in the message. So a simple way of doing it is just checking if they mentioned one person like so:

```py
def check(message):
    return message.author == ctx.author and len(message.mentions) == 1 and message.mentions[0] != ctx.author
```

This `check` predicate checks whether there was a mention and also makes sure that the person mentioned isn't the person using the command.

There is another issue in your second code and that's the handling of your `TimeoutError`, in Python this is a raised exception that must be caught to be handled, so we do something like this as seen in the [documentation](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html#discord.ext.commands.Bot.wait_for).

```py
try:
    msg = await client.wait_for('message', timeout=20, check=check)
except asyncio.TimeoutError:
    return await ctx.send("Cráááááá! What's the problem? Are you scared? Crááá! Call me when you stop crying out of fear! Cráááá!")
else:
    # In here, we didn't time out and msg is a valid discord.Message object
    opponent = msg.mentions[0]
```

Notice that I call `return` in the `except` clause so the function finishes when there is no valid opponent given.

After putting all this together you should have both a working program and an `opponent` variable pointing to the person you're playing against.

**Reply by Martyn0324** [reply]
Thanks, it worked! Just one more thing, though: I've seen that if I add `message.mentions[0] != client.user` to the `check` variable, the bot won't accept itself as a possible opponent. Is there a way to apply this filter to any other bots in a guild?
Also, I noticed that if I want to make the 2nd player choose an option, I can use the `opponent` variable. If I want to make the 1st player to choose right after that, should I still use `{ctx.author}` ?

**Reply by Rapptz** [reply]
Bots can be detected via [`User.bot`](https://discordpy.readthedocs.io/en/latest/api.html#discord.abc.User.bot). So in this case `message.author.bot` checks if the user is a bot.

You are correct that the original author is still `ctx.author`.
